Add support for Firebase Crashlytics

crayne opened the pull request:
<!-- Required: read https://github.com/Path-Check/covid-safe-paths/wiki/Pull-Request-Best-Practices for recommended best practices before opening your first pull request.  PR's raised not following those guidelines will require rework, so you might as well start off right -->

#### Description:

<!-- Description of what the PR does.  YOUR PR WILL BE REJECTED IF YOU DO NOT HAVE A DESCRIPTION -->
Added support for Firebase Crashlytics.  When gaen-mobile crashes, a stack trace will be generated in the Crashlytics console.  Crashlytics tells you what kind of device was involved.
#### Linked issues:

<!-- Add issues here e.g.: Fixes #1234 -->

#### Screenshots:

<!-- If you're changing visuals, add a screenshot here -->

#### How to test:

<!-- Description of how to validate or test this PR.  If it's a code change, you must describe what and how to test. -->
I tested this by dividing by zero in MainActivity.java.  I've found that there can be a lag of hours between the crash and when it shows up in the console.

The Firebase project name is "Pathcheck-GAEN-dev".  The nickname of the app within the project is "PathCheck - SafePaths GAEN (PC) - Android - Dev"

@octohub I have created a user id for you with owner role, using the email your Slack profile. Please let me know if you would prefer another email.

johnschoeman commented on android/app/build.gradle:
```diff
@@ -274,7 +280,7 @@ allprojects {
                 if (requested.group == 'com.google.firebase') {
                     // TODO verify no downstream issues with version bump
-                    details.useVersion '17+'
+                    //details.useVersion '17+'
```
should this be removed altogether?

crayne replied:
Yes

johnschoeman commented on android/app/google-services.json:
```diff
@@ -0,0 +1,40 @@
+      ],
+      "api_key": [
+        {
+          "current_key": "AIzaSyBVhlk_iGNcBzXGdWdeK1fIr-RZqkRgsso"
```
should we be saving all these keys in version control?  so should we add `android/app/google-services.json` to the .gitignore file?

crayne replied:
Makes sense to me

crayne replied:
The fetch is failing because I don't have the authority to do it.
The ios errors -- I don't know

juancruzgs commented on android/app/build.gradle:
```diff
@@ -197,13 +199,14 @@ android {
     flavorDimensions "version"
     productFlavors {
         bt {
-            applicationId 'org.pathcheck.covidsafepathsBt'
```
I think we were using this id to differentiate the GAEN app from the GPS one, and have both installed and the same time.
@johnschoeman @octohub We should confirm with id we are going to use before releasing versions to the Play Store because it can be changed afterwards.

crayne replied:
Can you suggest an alternative to this?

juancruzgs replied:
If possible I would keep `org.pathcheck.covidsafepathsBt`

crayne replied:
Did you mean change back to org.pathcheck.covidsafepathsBt in packages.json?  I tried that and it did not install.

juancruzgs replied:
I mean reverting the change here and in these 3 places
https://github.com/Path-Check/gaen-mobile/pull/71/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2L13
https://github.com/Path-Check/gaen-mobile/pull/71/files#diff-3e828e9281e5dfb593f53c53b1cec1d2L112
https://github.com/Path-Check/gaen-mobile/pull/71/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2L13

Did you get any error with the previous applicationId?

crayne replied:
Yes, that's why I changed them.

juancruzgs replied:
Do you know what error it was? 
Is it related to Firebase or it also happens on `develop` branch?

juancruzgs replied:
Oh I think the error is due to the uppercase letter
We can change the package name to `org.pathcheck.covidsafepaths.bt` in that case

crayne replied:
OK, i'll try that

crayne replied:
It only happened in the Firebase branch

juancruzgs commented:
@johnschoeman @octohub We should add one app for each ENV to Firebase, because they have different app ids and may have different functionalities in the future.
That's something we can do later, but it will be needed in order to get reports in production.

crayne commented:
@johnschoeman @juancruzgs Is there something I need to do before this can be merged?

aledustet commented:
@crayne sorry for the trouble but we think we might have fixed the issue with CI would you mind rebasing develop to kick the CI process again. Thank you so much!

juancruzgs commented on .gitignore:
```diff
@@ -103,3 +103,6 @@ safepaths.realm
 ios/COVIDSafePaths.xcworkspace/xcshareddata/swiftpm/Package.resolved
+
+# Firebase
+android/app/google-services.json
```
The app doesn't compile without a `google-services.json` file, right?
I can add it to the env configuration, just wanted to confirm if it's required

aledustet replied:
Yeah this sounds about right, is it possible to do something like the examples GoogleServices on the `ios` folder?

juancruzgs replied:
Sure!

crayne replied:
Yes, it is requred.  I will rebase.

crayne replied:
I got these error messages:
```
src/More/About.spec.tsx:8:8 - error TS2307: Cannot find module 'react-native-device-info' or its corresponding type declarations.

8 } from "react-native-device-info"
         ~~~~~~~~~~~~~~~~~~~~~~~~~~

src/More/About.tsx:15:8 - error TS2307: Cannot find module 'react-native-device-info' or its corresponding type declarations.

15 } from "react-native-device-info"
          ~~~~~~~~~~~~~~~~~~~~~~~~~~

```

juancruzgs replied:
Try running `yarn install`, I had the same error some minutes ago

juancruzgs replied:
We will need to commit a fake `google-services.json` to compile the app
This line won't be needed anymore

crayne commented:
Don't merge yet -- I have changed the Bt to bt

crayne commented:
I changed 
```org.pathcheck.covidsafepathsBt``` to ```org.pathcheck.covidsafepathsbt```
in package.json, Fastfile and the app-level build.gradle.
I got the following error:
```
> Task :app:processBtDebugGoogleServices FAILED
Parsing json file: /Users/susancrayne/path-check-gaen-mobile-fork/gaen-mobile/android/app/google-services.json

```
so I reverted the changes and pushed the branch again after merging from upstream

juancruzgs commented:
Have you updated the appId in `google-services.json`?
That might be the issue

crayne commented:
I'm afraid that won't work. ```google-services.json``` is generated by Firebase to go with the app configuration, and the app id in Firebase is ```org.pathcheck.covidsafepaths```

crayne commented:
There is a way to have a second ```google-services.json``` associated with your build variant:
```
https://developers.google.com/android/guides/google-services-plugin
```
The second file would have the "bt" suffix on the package name.
However, I don't know where I would place such a file in the project structure.

crayne commented:
Definitely

crayne commented:
@juancruzgs @johnschoeman I changed the applicationId to ```org.pathcheck.covidsafepaths.bt```, and created an app with the new name in Firebase . I believe that this is ready to merge but can't be merged because of the failing tests? If so, @johnschoeman  suggested a solution above:
```
Let me know if this is blocking it might make sense to add you as a collaborator if we can resolve this quickly.

```

juancruzgs commented:
The CI environment needs a `google-services.json` to compile the app
Could you add this sample `google-services.json`? I removed all the sensitive information
```
{
  "project_info": {
    "project_number": "",
    "project_id": ""
  },
  "client": [
    {
      "client_info": {
        "mobilesdk_app_id": "1:123456789012:android:1234567890123456",
        "android_client_info": {
          "package_name": "org.pathcheck.covidsafepaths.bt"
        }
      },
      "oauth_client": [
        {
          "client_id": "",
          "client_type": 3
        }
      ],
      "api_key": [
        {
          "current_key": ""
        }
      ],
      "services": {
        "appinvite_service": {
          "other_platform_oauth_client": []
        }
      }
    }
  ],
  "configuration_version": "1"
}
```

crayne commented:
I'll try, but I'm not sure that Firebase is going to like it without the things you removed.

juancruzgs commented:
I tested it in your branch with the json I posted and it worked, but let me know if you run into any issues

crayne commented:
Firebase does like it.  Thanks.

crayne commented:
I'm looking into the remaining error.

crayne commented:
You will notice that ``` Run pod install --repo-update``` has passed now.  Here is what I did:
```
rvm install 2.7.1
rvm use 2.7.1
move Gemfile from ios directory up one level
bundle
yarn install
sudo gem install cocoapods
pod update Firebase/Crashlytics
yarn install:pod
```
These steps were sufficient but not all necessarily necessary.  I will now try to see why the next step is failing

crayne commented:
@juancruzgs  The test that is failing is 
```
  Run xcodebuild test -workspace COVIDSafePaths.xcworkspace \
```
I have successfully built and run the iOS version on a simulator (iPhone 11)
This is the command line for the test:
```
    /Applications/Xcode_11.5.app/Contents/Developer/usr/bin/xcodebuild test -workspace 
COVIDSafePaths.xcworkspace -scheme BT_Development -destination "platform=iOS Simulator,name=iPhone 8"
```
Please note that the simulator specified in iPhone 8.  At any rate, since the app builds and runs on iOS, there must be an issue with the test.  My knowledge of test scripting is zero, so I will bounce this back to you.

juancruzgs commented:
@mattThousand iOS unit tests are failing in this branch, do you know what could be the issue?

crayne commented:
@juancruzgs I ran the test from the command line and got this error:
```
Testing failed:
	Could not get GOOGLE_APP_ID in Google Services file from build environment
	Testing cancelled because the build failed.

** TEST FAILED **
```
This leads me to believe I need to set up a separate app in Firebase for the iOS version.  In the best of all possible worlds the test would know that there is no iOS Firebase configuration and not fail, but since the iOS piece is needed anyway I can do it.

juancruzgs commented:
We do have a `GoogleService-Info.plist` file in the repository
https://github.com/Path-Check/gaen-mobile/blob/develop/ios/GoogleService-Info.plist
I'm not sure if it's already set up

crayne commented:
I think you need to download it from Firebase for each app you create.  I can do this all tomorrow.

octohub commented:
@crayne There was a change made to the build checks so you'll have to merge the latest from `develop` and push to this branch so the `build_debug_android` passes.  Sorry for the hassle!

crayne commented:
Ok

crayne commented:
@johnschoeman  @juancruzgs @aledustet I installed Firebase/Crashlytics for the iOS app, and that made all the checks pass.  I'm not sure that Crashlytics works for iOS (it certainly works for Android).  What I'd like to do is merge this and then make a separate issue to work on Crashlytics for iOS.  That way, Crashlytics will be available for Android development right away.

crayne commented:
more work on Crashlytics for iOS

aledustet commented on ios/Podfile:
```diff
@@ -60,7 +60,9 @@ def bt_pods
   permissions_path = '../node_modules/react-native-permissions/ios'
   pod 'Alamofire', '~> 4.9.1'
   pod 'ZIPFoundation'
+  pod 'Firebase/Analytics'
```
I'm not sure if we are supposed to include this package for security concerns, @mattThousand thoughts?

mattThousand replied:
Crashlytics for iOS was set up in #11 actually 😄  - this scope of this PR should be for Android

aledustet commented on vendor/bundle/ruby/2.7.0/gems/simctl-1.6.8/spec/SampleApp/SampleApp.xcodeproj/project.pbxproj:
```diff
@@ -0,0 +1,269 @@
+// !$*UTF8*$!
```
Should this files be included in the commit?

aledustet commented on ios/AppDelegate.m:
```diff
@@ -52,6 +54,8 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
 
   [RNSplashScreen showSplash:@"LaunchScreen" inRootView:rootView];
+  //Crashlytics test
+  //assert(NO);
```
How do we feel about removing this commented code? If this is something we will need in the future we can circle back to the implementation or the docs setting up the library?

aledustet commented on ios/GoogleService-Info.plist:
```diff
@@ -1,40 +1,36 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
 <plist version="1.0">
-  <dict>
```
I'm not sure we want to keep this file in the repo, we have an example.GoogleService-Info.plist in the ios folder that a script that sets it up is copying on CI. I'm creating a ticket to update the docs with this info.

mattThousand commented on ios/AppDelegate.m:
```diff
@@ -21,6 +21,10 @@
 @implementation AppDelegate
 - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
```
Crashlytics was set up in #11 , so changes here can be removed

crayne replied:
It didn't look like Crashlytics was set up for iOS in this repo because the iOS apps in the Pathcheck-GAEN-dev project in Firebase are not enabled for Crashlytics.  Now that I look at  #11 I see I was wrong. I went down this path to get a GoogleService-info.plist to copy because the error in the CI test indicated something wrong with that.  If the file is being copied in the fetch-ha script, I can't run that script because I don't have permission.  I will back out the changes and see if the problem can be solved another way.

mattThousand commented on ios/COVIDSafePaths.xcodeproj/project.pbxproj:
```diff
@@ -683,6 +683,8 @@
 				B5D968F324C6036A008DE452 /* Crashlytics */,
 				B53575E124D0C2DE00E7310D /* Upload Symbols */,
+				EDB42E500C5E02AE9793F103 /* [CP-User] [RNFB] Core Configuration */,
+				E1BAB01724D499A10075B019 /* ShellScript */,
```
Here as well

mattThousand commented on ios/Podfile:
```diff
@@ -60,7 +60,9 @@ def bt_pods
+  pod 'Firebase/Analytics'
   pod 'Firebase/Crashlytics'
+  pod 'RNFBCrashlytics', :path => '../node_modules/@react-native-firebase/crashlytics'
 end
```
Here as well

mattThousand commented on ios/Podfile.lock:
```diff
@@ -12,36 +12,45 @@ PODS:
-  - Firebase/Crashlytics (6.25.0):
+  - Firebase/Analytics (6.27.1):
+    - Firebase/Core
+  - Firebase/Core (6.27.1):
```
Here as well

mattThousand commented on vendor/bundle/ruby/2.7.0/gems/terminal-notifier-2.0.0/vendor/terminal-notifier/Terminal Notifier.xcodeproj/project.pbxproj:
```diff
@@ -0,0 +1,315 @@
+// !$*UTF8*$!
+{
+	archiveVersion = 1;
```
Here as well

mattThousand commented on vendor/bundle/ruby/2.7.0/gems/simctl-1.6.8/spec/SampleApp/SampleApp.xcodeproj/project.pbxproj:
```diff
@@ -0,0 +1,269 @@
+// !$*UTF8*$!
+{
+	archiveVersion = 1;
+	classes = {
```
Here as well

crayne commented:
I will make a new branch
